lib/smart_answer/state_resolver.rb

```diff
@@ -0,0 +1,85 @@
+module SmartAnswer
+  class StateResolver
+    def initialize(flow)
+      @flow = flow
+    end
+
+    def state_from_response_store(response_store, requested_node = nil)
+      forwarding_responses = @flow.response_store == :session ? {} : response_store.all
+      establish_response_store_state(start_state(forwarding_responses),
+                                     response_store,
+                                     requested_node)
+    end
+
+    def state_from_params(params)
+      responses = params[:responses].to_s.split("/")
+
+      state = responses.inject(start_state) do |current_state, response|
+        return current_state if current_state.error
+
+        transition_state_to_next_node(current_state, response)
+      end
+
+      if params[:next]
+        transition_state_to_next_node(state, params[:response])
+      elsif params[:previous_response]
+        apply_response_to_state(state, params[:previous_response])
+      else
+        state
+      end
+    end
+
+  private
+
+    def establish_response_store_state(state, response_store, requested_node)
+      node_name = state.current_node.to_s
+      no_response_for_question = !response_store.has?(node_name)
+      reached_an_outcome = @flow.node(node_name.to_sym).outcome?
+
+      return state if no_response_for_question || reached_an_outcome || state.error
+
+      response = response_store.get(node_name)
```

Comment:
Nit: I feel like this could be made more concise, whilst it's less verbose - it's easier to read because there's fewer lines/variables to follow.

```suggestion
      response = response_store.get(node_name)
      reached_an_outcome = @flow.node(node_name.to_sym).outcome?

      return state if response.nil? || reached_an_outcome || state.error
```

Reply:
Thanks - amended.